app/src/main/kotlin/compiler/intermediate_form/ControlFlow.kt

```diff
@@ -9,17 +9,284 @@ import compiler.ast.StatementBlock
+                        firstStageRecursion(astNode.leftOperand, leftModifiedVariables)
+
+                        val rightModifiedVariables = ReferenceHashSet<Variable>()
+                        firstStageRecursion(astNode.rightOperand, rightModifiedVariables)
+
+                        modifiedUnderCurrentBase.addAll(leftModifiedVariables)
+                        modifiedUnderCurrentBase.addAll(rightModifiedVariables)
+                        invalidatedVariables[astNode] = rightModifiedVariables
+                    }
+                    else {
+                        firstStageRecursion(astNode.rightOperand, modifiedUnderCurrentBase)
+                        firstStageRecursion(astNode.leftOperand, modifiedUnderCurrentBase)
+                    }
+                }
+
+                is Expression.FunctionCall -> {
+                    val variablesModifiedInArguments = ReferenceHashSet<Variable>()
+                    for (argumentNode in astNode.arguments) {
+                        val modifiedVariables = ReferenceHashSet<Variable>()
+                        firstStageRecursion(argumentNode.value, modifiedVariables)
+                        variablesModifiedInArguments.addAll(modifiedVariables)
+                    }
+
+                    val variablesModifiedByCall = getVariablesModifiedBy(nameResolution[astNode] as Function)
+                    invalidatedVariables[astNode] = variablesModifiedByCall
+                    modifiedUnderCurrentBase.addAll(variablesModifiedByCall)
+                    modifiedUnderCurrentBase.addAll(variablesModifiedInArguments)
+                }
+
+                is Expression.Conditional -> {
+                    val conditionModifiedVariables = ReferenceHashSet<Variable>()
+                    firstStageRecursion(astNode.condition, conditionModifiedVariables)
+
+                    val trueBranchModifiedVariables = ReferenceHashSet<Variable>()
+                    firstStageRecursion(astNode.resultWhenTrue, trueBranchModifiedVariables)
+
+                    val falseBranchModifiedVariables = ReferenceHashSet<Variable>()
+                    firstStageRecursion(astNode.resultWhenFalse, falseBranchModifiedVariables)
+
+                    modifiedUnderCurrentBase.addAll(conditionModifiedVariables)
+                    modifiedUnderCurrentBase.addAll(trueBranchModifiedVariables)
+                    modifiedUnderCurrentBase.addAll(falseBranchModifiedVariables)
+
+                    invalidatedVariables[astNode] = combineReferenceSets(trueBranchModifiedVariables, falseBranchModifiedVariables)
+                }
+            }
+        }
+
+        firstStageRecursion(expression, ReferenceHashSet())
+
+        // second stage is to actually produce CFG
+        var currentTemporaryRegisters = ReferenceHashMap<Variable, Register>()
+
+        fun mergeCFGsUnconditionally(first: ControlFlowGraph?, second: ControlFlowGraph?): ControlFlowGraph? {
```

Comment:
There's a [builder class](https://github.com/TCS-Compilers/compiler/blob/function-details-generator-generate-call/app/src/main/kotlin/compiler/intermediate_form/ControlFlowGraphBuilder.kt) performing very similar operations being implemented on a parallel branch (which will be merged soon). Just letting you know, as this logic (and the method below)  can probably be easily moved there.